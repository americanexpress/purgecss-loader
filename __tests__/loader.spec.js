/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

const path = require('path');
const webpack = require('webpack');
const MemoryFs = require('memory-fs');
const findIndex = require('lodash/findIndex');

jest.setTimeout(10000);

const runLoader = ({
  entry = '../__fixtures__/Component.jsx',
} = {}) => {
  const compiler = webpack({
    context: __dirname,
    entry,
    output: {
      path: path.resolve(__dirname),
    },
    module: {
      rules: [
        {
          test: /\.jsx?$/,
          use: [{
            loader: 'babel-loader',
            options: {
              presets: [
                'babel-preset-env',
                'babel-preset-react',
              ],
            },
          }],
        },
        {
          test: /\.s?css$/,
          use: [
            { loader: 'style-loader' },
            {
              loader: 'css-loader',
              options: {
                modules: true,
                localIdentName: '[name]__[local]___[hash:base64:5]',
              },
            },
            {
              loader: path.resolve(__dirname, '../loader.js'),
              options: {
                paths: [path.resolve(__dirname, '../__fixtures__/**/*.{js,jsx}')],
                whitelistPatternsChildren: [/:global$/],
              },
            },
          ],
        },
      ],
    },
    externals: {
      react: {
        var: 'React',
        commonjs2: 'react',
      },
    },
  });

  compiler.outputFileSystem = new MemoryFs();

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err);

      resolve(stats);
    });
  });
};

describe('purgecss loader', () => {
  it('should strip unused classes', () => {
    expect.assertions(2);
    return runLoader()
      .then((stats) => {
        const { modules } = stats.toJson();
        const cssModuleIndex = findIndex(modules, {
          name: '../node_modules/css-loader??ref--5-1!../loader.js??ref--5-2!../__fixtures__/styles.css',
        });
        const output = modules[cssModuleIndex].source;
        expect(output).not.toContain('isNotUsed');
        expect(output).toContain('isUsed');
      });
  });

  it('should not strip global classes', () => {
    expect.assertions(2);
    const componentEntry = { entry: '../__fixtures__/ComponentGlobal.jsx' };
    return runLoader(componentEntry)
      .then((stats) => {
        const { modules } = stats.toJson();
        const cssModuleIndex = findIndex(modules, {
          name: '../node_modules/css-loader??ref--5-1!../loader.js??ref--5-2!../__fixtures__/root.scss',
        });
        const output = modules[cssModuleIndex].source;
        expect(output).not.toContain('isNotUsed');
        expect(output).toContain('isUsed');
      });
  });
});
