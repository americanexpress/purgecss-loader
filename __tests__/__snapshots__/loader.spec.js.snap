// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`purgecss loader should strip unused classes 1`] = `
"exports = module.exports = require(\\"../node_modules/css-loader/lib/css-base.js\\")(false);
// imports


// module
exports.push([module.id, \\"/*\\\\n * Copyright (c) 2018 American Express Travel Related Services Company, Inc.\\\\n *\\\\n * Licensed under the Apache License, Version 2.0 (the \\\\\\"License\\\\\\"); you may not use this file except\\\\n * in compliance with the License. You may obtain a copy of the License at\\\\n *\\\\n * http://www.apache.org/licenses/LICENSE-2.0\\\\n *\\\\n * Unless required by applicable law or agreed to in writing, software distributed under the License\\\\n * is distributed on an \\\\\\"AS IS\\\\\\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express\\\\n * or implied. See the License for the specific language governing permissions and limitations under\\\\n * the License.\\\\n */\\\\n\\\\n.styles__isUsed___3xJqg {\\\\n  color: red;\\\\n}\\\\n\\", \\"\\"]);

// exports
exports.locals = {
	\\"isUsed\\": \\"styles__isUsed___3xJqg\\"
};"
`;
